bauhaus/constraint_builder.py

```diff
@@ -128,40 +138,66 @@ def build(self, propositions) -> 'NNF':
             A built NNF constraint
 
         """
+        if self._groupby:
+            assert len(propositions) == 1, "Should only be using groupby on a single class"
+
+        def partition(inputs):
+            if not self._groupby:
+                return [inputs]
+            elif isinstance(self._groupby, str):
+                partitions = {}
+                for var in inputs:
+                    val = getattr(var.name, self._groupby)
+                    if val not in partitions:
+                        partitions[val] = []
+                    partitions[val].append(var)
+                return partitions.values()
+            else:
+                return self._groupby(inputs)
+
         if self._constraint is _ConstraintBuilder.implies_all:
             left_vars = unpack(self._left, propositions) if self._left else []
             right_vars = unpack(self._right, propositions) if self._right else []
             if self._func:
                 # retrieve dictionary of inputs
-                inputs = self.get_implication_inputs(propositions) 
+                inputs = self.get_implication_inputs(propositions)
                 if not any(inputs.values()) and not right_vars:
                     raise ValueError(f"The '{self}' cannot be built"
                                      " as it is decorating a class and"
                                      " the right implication variables are not"
                                      " provided. If it is decorating a method,"
-                                     " ensure that the method's return is" 
+                                     " ensure that the method's return is"
                                      " valid for bauhaus or for the nnf library."
                                      " Check your decorator signature and set"
                                      " the 'right' keyword argument to such a value.")
             else:
                 # inputs should be empty if not created from a decorator
                 inputs = []
-            return self._constraint(self, inputs, left_vars, right_vars)
-        
+
+            constraints = []
+            for input_set in partition(inputs):
+                constraints.append(self._constraint(self, input_set, left_vars, right_vars))
+            return And(constraints)
+
         inputs = self.get_inputs(propositions)
         if not inputs:
             raise ValueError(inputs)
-        if self._constraint is _ConstraintBuilder.at_most_k:
-            return self._constraint(self, inputs, k=self._k)
-        return self._constraint(self, inputs)
+
+        constraints = []
+        for input_set in partition(inputs):
+            if self._constraint is _ConstraintBuilder.at_most_k:
+                constraints.append(self._constraint(self, input_set, k=self._k))
```

Comment:
k isn't a kwarg, so the func signature would be `self._constraint(self, input_set, self._k)`